core/src/avm2/globals/flash/xml/XMLDocument.as

```
@@ -1,21 +1,84 @@
+         // `<!-- comment -->` gets parsed as a comment
+         // ` foo` is another text node
+         // To achieve this, just wrap it all together in a parent.
+
+         var oldSettings = XML.AS3::settings();
+         var newSettings = XML.AS3::defaultSettings();
+         newSettings.ignoreWhitespace = this.ignoreWhite;
+         XML.AS3::setSettings(newSettings);
+
+         try {
+            clear();
+            var root = new XML("<xml>" + input + "</xml>");
+            for each (var child in root.children()) {
+               var node = _convertXmlNode(child);
+               if (node != null) {
+                  appendChild(node);
+               }
+            }
+         } finally {
+            XML.AS3::setSettings(oldSettings);
+         }
+      }
+
+      private function _convertXmlNode(original: XML): XMLNode {
+         var nodeType = _convertXmlNodeType(original.nodeKind());
+         if (nodeType == 0) {
+            return null;
+         }
+         var nodeValue = nodeType == XMLNodeType.ELEMENT_NODE ? original.name() : original.toString();
+         var result = new XMLNode(nodeType, nodeValue);
+         for each (var originalChild in original.children()) {
+            var child = _convertXmlNode(originalChild);
+            if (child != null) {
+               result.appendChild(child);
+            }
+         }
+         var attributeList = original.attributes();
+         var attributes = {};
+         for each (var attribute in attributeList) {
+            attributes[attribute.name()] = attribute.toString();
+         }
+         result.attributes = attributes;
+         return result;
+      }
+
+      private function _convertXmlNodeType(kind: String): uint {
+         if (kind == "text") {
+            return XMLNodeType.TEXT_NODE;
+         }
+         if (kind == "comment") {
+            return XMLNodeType.COMMENT_NODE;
+         }
+         if (kind == "processing-instruction") {
```

I don't think this is possible. We don't allow processing-instructions during parsing. 

In general: I would  suggest we just throw for what I think is the impossible situation of an unexpected kind here. (Would also eliminate some null checks above)

---

We don't exactly have a way to panic or throw RustErrors during AS, and throwing a new type of AS error is weird and unprecedented.

---

I don't think it's unreasonable to do `throw new Error("unexpected state")` for something that we believe is impossible. This seems a lot easier to debug later than the number of children being different or something like that.

---

Could you add in a stub, so that there's some indication if this ever gets hit?

---

Stub will incorrectly mark this as a stub, which it isn't. I'll add an error.